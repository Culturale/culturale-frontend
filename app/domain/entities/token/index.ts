export * from './token';
export * from './token.interface';
