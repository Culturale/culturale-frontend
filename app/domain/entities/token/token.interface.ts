export interface RefreshTokenResponse {
  accessToken: string;
  expiresIn: number;
  idToken?: string;
  refreshToken?: string;
  scope?: string;
  tokenType: string;
}

export interface IToken {
  accessToken: string;
  refreshToken?: string;
  invalid?: boolean;

  readonly userId: string;
  invalidate: () => void;
  update: (credentials: RefreshTokenResponse) => void;
}

export interface IFleetEngineAccessToken {
  expirationAt: number;
  token: string;
  readonly millisecondsToExpire: number;
}
