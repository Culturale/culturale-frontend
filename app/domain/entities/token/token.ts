import jwtDecode from 'jwt-decode';
import { action, computed, makeObservable, observable } from 'mobx';

import type { IToken, RefreshTokenResponse } from './token.interface';

export class Token implements IToken {
  public accessToken: string;
  public refreshToken?: string;
  public invalid?: boolean;

  constructor(accessToken: string, refreshToken?: string, invalid?: boolean) {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.invalid = invalid;

    makeObservable(this, {
      accessToken: observable,
      invalid: observable,
      invalidate: action,
      refreshToken: observable,
      update: action,
      userId: computed
    });
  }

  public get userId(): string {
    const decodedToken = jwtDecode(this.accessToken);
    return decodedToken && decodedToken['https://paack.co'].id;
  }

  public invalidate() {
    this.invalid = true;
  }

  public update(credentials: RefreshTokenResponse) {
    this.accessToken = credentials.accessToken;

    if (credentials.refreshToken) {
      this.refreshToken = credentials.refreshToken;
    }
  }
}
