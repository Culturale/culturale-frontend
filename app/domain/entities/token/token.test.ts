import { Token } from './token';
import type { IToken } from './token.interface';

describe('Token', function () {
  let instance: IToken;

  beforeEach(function () {
    instance = new Token('access-token', 'refresh-token', false);
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });

  it('invalidate sets valid as false', function () {
    instance.invalidate();

    expect(instance.invalid).toBeTruthy();
  });
});
