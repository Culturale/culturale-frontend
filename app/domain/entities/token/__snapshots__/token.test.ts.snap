// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token can be created 1`] = `
Token {
  "accessToken": "access-token",
  "invalid": false,
  "refreshToken": "refresh-token",
}
`;
